### PR TITLE
feat: allow prefix for any generative model

### PR DESCRIPTION
**Pull request by borisdayma**
Allow `padding_text` for any generative model.

### Motivations

* some models require some text at the beginning such as GPT-2 which should always have a white space. GPT-2 model could now have `config.padding_text` set to `" "` to ensure this is done
* this argument adds more customization to huggingface inference widget (related issue #5553). For huggingtweets models I can set `model.config.task_specific_params['text-generation']['padding_text'] = '<|endoftext|> '`

See [example](https://gist.github.com/borisdayma/89aaf1587390340add44c3c7081bffcd) of use of `padding_text` to create negative/positive sentences.

*Side note*: this is not completely obvious that a leading space should always be added for GPT-2 as `" \n This"` is not tokenized the same as `" \nThis"` which GPT-2 may have seen more in its training data (unless white spaces are added after new lines as well)

### Future improvements

In the future, we should also add `prefix` and `suffix`. This could let us do things similar to GPT-3 with the inference widget and specific tasks:
* `padding_text` would be used to initialize tasks (ex: 10 examples conversation with a bot). This is always cropped out and never displayed in the output
* `prefix` would add required text (ex: `" Human: "`). This is displayed in output.
* `suffix` would add required text after prompt (ex: `" \n AI: "`). This is displayed in output.

### CODE REVIEW

**Comment by borisdayma** [comment]
Another solution (though I think not as clean), would be to allow `prefix` element like in `TranslationPipeline`, though it is not as flexible (as it will become part of input and always display).

As a matter of fact, `TranslationPipeline`, `SummarizationPipeline` and `TextGenerationPipeline` are very similar (may benefit from some refactoring)!
As a temporary solution for my specific case, I could turn my GPT-2 into a translation model. Not sure it would trick the inference widget as I understand it detects first the model architecture to define which task to use… Also it's very hacky…

**Comment by borisdayma** [comment]
I updated my [example](https://gist.github.com/borisdayma/89aaf1587390340add44c3c7081bffcd) of the feature with the case when it loads model config.

**Comment by borisdayma** [comment]
Let me know if I can provide any additional information or make any changes

**Comment by borisdayma** [comment]
Just checking if you need anything else from me @julien-c

**Comment by julien-c** [comment]
I’m off for a couple of days so I’ll let others chime in

**Comment by borisdayma** [comment]
Enjoy your time off!

**Comment by borisdayma** [comment]
@julien-c @mfuntowicz sorry I'm just pinging.
I have a [demo](https://huggingface.co/huggingtweets/julien_c) I'm excited about and this feature would make it so much better.

**Comment by julien-c** [comment]
Thoughts @patrickvonplaten?

**Comment by julien-c** [comment]
Thoughts @patrickvonplaten @sshleifer? (text generation pipeline option)

File: examples/text-generation/run_generation.py
Comment (sshleifer) [inline examples/text-generation/run_generation.py:183]: why isn't this called prefix?

**Review by sshleifer — APPROVED** [review]
LGTM besides the naming. To me padding means the dummy tokens we add to make our input_ids into rectangular tensors, and this seems like it should just be called prefix.
That said, I understand the naming was there before this PR and am happy to merge as is!

File: examples/text-generation/run_generation.py
Comment (sshleifer) [reply examples/text-generation/run_generation.py:183]: Were we just ignoring this arg before?

File: src/transformers/pipelines.py
Comment (sshleifer) [inline src/transformers/pipelines.py:692]:
```suggestion
                if generate_kwargs.get("max_length", None) is not None:
```

File: examples/text-generation/run_generation.py
Comment (borisdayma) [reply examples/text-generation/run_generation.py:183]: I agree that it should be renamed

**Comment by borisdayma** [comment]
This padding adds the script functionality to pipelines (and allow for custom text).

However the main idea would be to eventually allow "GPT-3 style" API's.

image

It could be renamed "prefix" but I suggest to make a distinction between padding and prefix where you would potentially want to display prefix & suffix in an interface (for example huggingface widget) but not necessarily the "padding".

This kind of API could be used for T5 and can let people create more modular inference examples.

File: src/transformers/pipelines.py
Comment (borisdayma) [reply src/transformers/pipelines.py]: Nice!

**Comment by borisdayma** [comment]
@sshleifer I gave more thoughts to it and I agree we should just name it `prefix` and maybe add a "deprecated warning" on `run_generation`.
It should be sufficient for now and we will always be able to add a more complex behavior in the future if necessary.

**Comment by sshleifer** [comment]
style: you need to rebase and then `pip install black --upgrade`
Other CI failures are spurious, you can ignore them.

My last request would be to add test coverage for the new argument in test_pipelines.py.
see:
```
test_torch_text_generation
test_tf_text_generation
```

**Comment by borisdayma** [comment]
Changes:

* renamed `padding_text` to `prefix` in pipelines and `run_generation.py` (with warning of deprecated argument)
* since PR was made, we cannot tokenize empty text anymore so we now check for empty "prefix"
* since PR was made, we cannot pass extra args to `generate` so we explicitly add `prefix` to pipeline `__call__` method
* added tests for pipeline using prefix

Note:

* we set default `prefix` to `None` (vs `""`) as we could pass `""` to overwrite model default value.

File: src/transformers/pipelines.py
Comment (sshleifer) [inline src/transformers/pipelines.py:866]: same .get logic works on this line

**Review by sshleifer — APPROVED** [review]

File: src/transformers/pipelines.py
Comment (sshleifer) [inline src/transformers/pipelines.py:858]: self.PREFIX -> self.XL_PREFIX if this is the only place it is used.

**Review by sshleifer — APPROVED** [review]
LGTM

File: src/transformers/pipelines.py
Comment (sgugger) [inline src/transformers/pipelines.py:830]:
```suggestion
            prefix (:obj:`str`, `optional`):
```
(Nit: no need to mention the defaults to None for something that is optional.)

**Review by sgugger — APPROVED** [review]
Thanks for your work on this!

**Review by LysandreJik — APPROVED** [review]
This is a welcome change, thanks for the deprecation. **Will merge this after the 3.1.0 release.**

Regarding your comment @borisdayma I don't think GPT-2 requires a space before tokens. Not having a space means it is the initial token beginning the sequence, which is useful information, especially as the model was trained like this.

**Comment by borisdayma** [comment]
Thanks @LysandreJik. Yes I noticed later that my results were better without the initial space as well.

Intuitively it seemed to me like a waste of encoding differentiating between having a space or not (for example even after new lines) as the start of a document can simply be identified by the special token.
I thought there were more examples without the space within the original training set which would help my fine-tuning, as long as I do the same both for training and inference.
In the end, I get my best results with just `<|endoftext|>` between each sample and stripping every extra space.

**Comment by borisdayma** [comment]
@LysandreJik let me know if anything else is required on my side.
Since it's a relatively large PR it can be hard to remain in sync with master.

**Review by patrickvonplaten — APPROVED** [review]

**Comment by patrickvonplaten** [comment]
I think we can merge this: @LysandreJik I leave it to you to merge :-)

**Comment by julien-c** [comment]
+ @mfuntowicz can you update the transformers dependency in the inference API when this lands on master, so that @borisdayma can showcase his magic ;)

**Comment by borisdayma** [comment]
@mfuntowicz can you let me know when you update the inference API so I do some tests on the widget?

**Comment by mfuntowicz** [comment]
@borisdayma Everything has been deployed on the inference API, servers are up. 

Thanks for taking care of this 💪 🙏

**Comment by borisdayma** [comment]
That was fast! Thanks it looks to be working great!

**Pull request merged**
